### DISCUSSION TITLE
Docs/SQL: Why is the db Session injected into the endpoints, not in the CRUD-operations itself?

### DISCUSSION THREAD

**Discussion by axeljaeger-arculus** [fastapi/fastapi#9523] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
The example is visible here:
https://fastapi.tiangolo.com/tutorial/sql-databases/#create-the-database-tables
```

### Description

I wonder why the example for using an SQL database with FastAPI injects the db Session into the REST endpoints and these then only pass it down to the CRUD operations. Why is it recommended or a good idea to not inject the db session directly in the CRUD methods?

### Operating System

macOS

### FastAPI Version

0.67

### Python Version

3.11

**Comment by acidjunk** [comment]
When you have an app that uses DB on every endpoint: I think it's perfectly fine to init the session globally.

Example:
https://github.com/acidjunk/pricelist-fastapi

But, for the record, I am also interested in good reasons for the approach in the docs.

**Reply by BredoGen** [reply]
I guess @axeljaeger-arculus means the dependency injection of db session in the service layer (crud).

I'm also interested why the db session is injected in the endpoint, while we can do something like this:

```python
@router.get("/", response_class=UserListSchema)
async def list_users(service: dependencies.UserServiceDep):
    return await service.list_all()
```

```python
def get_user_service(db_session: DBSessionDep) -> UserService:
    return UserService(db_session)

UserServiceDep = Annotated[UserService, Depends(get_user_service)]
```
This way we have isolated service layer which receives the db session and http endpoint doesn't know about service's deps.

Going deeper it's also a good idea to use app.dependency_overrides for easy deps mock.
